cabby/data/metagraph/utils.py

```diff
@@ -103,11 +108,88 @@ def update_osm_map(osm_map: Map,
+        name = poi_row["brand"]
+    elif isinstance(poi_row["shop"], str):
+        name = poi_row["shop"]
+    elif isinstance(poi_row["tourism"], str):
+        name = poi_row["tourism"]
+    elif isinstance(poi_row["amenity"], str):
+        name = poi_row["amenity"]
+    return "%s_%s" % (name, poi_row["osmid"])
+
+def convert_multidi_to_weighted_undir_graph(
+  in_graph: nx.MultiDiGraph) -> nx.Graph:
+  out_graph = nx.Graph()
+  for u, v in in_graph.edges():
+      if out_graph.has_edge(u,v):
+          out_graph[u][v]['weight'] += 1.0
+      else:
+          out_graph.add_edge(u, v, weight=1.0)
+  return out_graph
+
 def construct_metagraph(region: Region,
                         s2_level: int,
-                        base_osm_map_filepath: Text):
+                        s2_node_levels: Sequence[int],
+                        base_osm_map_filepath: Text) -> nx.Graph:
+  # Get relation data and add to existing graph.
   wd_relations = query.get_geofenced_wikidata_relations(region,
                                                         extract_qids=True)
   osm_map = Map(region, s2_level, base_osm_map_filepath)
   update_osm_map(osm_map, wd_relations)
-  return osm_map, wd_relations
\ No newline at end of file
+
+  # Construct initial metagraph and human-readable node names.
+  metagraph = convert_multidi_to_weighted_undir_graph(osm_map.nx_graph)
+  osmid_to_name = {}
+  name_to_point = {}
+  name_to_wikidata = {}
+  for _, row in osm_map.poi.iterrows():
+    name = construct_human_readable_name(row)
+    assert isinstance(name, str)
+    osmid_to_name = {row["osmid"]: name}
+    name_to_point = {name: row["geometry"]}
+    if isinstance(row["wikidata"], str):
+      name_to_wikidata[name] = row["wikidata"]
+  nx.relabel.relabel_nodes(metagraph, osmid_to_name,
+                           copy=False)
+
+  # Set useful node attributes.
+  name_to_osmid = {osmid: name for name, osmid in osmid_to_name.items()}
+  nx.set_node_attributes(metagraph, name_to_osmid, name="osmid")
+  nx.set_node_attributes(metagraph, name_to_point, name="geometry")
+  nx.set_node_attributes(metagraph, name_to_wikidata, name="wikidata")
+
+  # Add conceptual edges from wd_relations.
+  wikidata_to_name = {wikidata: name for
+                      name, wikidata in name_to_wikidata.items()}
+  for _, row in wd_relations.iterrows():
+    place_node_name = wikidata_to_name[row["place"]]
+    concept_node_name = "%s_%s" % (row["instanceLabel"], row["instance"])
+    metagraph.add_edge(place_node_name, concept_node_name, weight=1.0)
+    metagraph.add_edge(concept_node_name, place_node_name, weight=1.0)
+
+  # S2 cell nodes and edges.
+  for node, data in metagraph.nodes.data():
+    if not data or "geometry" not in data:
+      continue
+    geometry = data["geometry"]
+    for level in s2_node_levels:
+      s2_cell_id = util.cellid_from_point(geometry, level)
+      metagraph.add_edge(node, s2_cell_id)
+      metagraph.add_edge(s2_cell_id, node)
```

Comment:
Don't you just need to add this once since it is an undirected graph?

Reply:
Fixed by 32fbff65bfb0fb88b677bbec884b3583ba534b56